Default the COP's cop_features to the bitmask of the default-on featu…

demerphq opened the pull request:
…res.

The assumption is that cop_features holds the bitmask of the currently enabled feature set, and hence that S_magic_sethint_feature() can incrementally update it as keys for features are set in/cleared from %^H.

However, this overlooks one problem - the default set of features enabled (keys) is not empty, the default for %^H *is* empty, but as long as %^H is empty the internals knows that it is *not* running with a custom feature set, so avoids the complexity of setting up %^H.

There's a C test for features that knows about this, but it was only correctly handling (Perl) compile-time tests for features. That hasn't been noticed before, because all the "default on" features are compile-time only.

This commit fixes things so that run-time default-on features also work properly, by setting each COP's cop_features to the value of the bits set at the COP's compile time - ie the same approach as cop_hints.

tonycoz committed:
op.c - copy cop_features to the COP inserted into the OP tree

All current features are compile-time, so this hasn't been an issue,
but my current implementation of the autovivification issue checks
features at runtime, which failed to work correctly if non-bundle
features were set.

This commit fixes things so that run-time default-on features also work
properly, by setting each COP's cop_features to the value of the bits
set at the COP's compile time - ie the same approach as cop_hints.

leonerd commented on regen/feature.pl:
```diff
@@ -302,6 +302,8 @@ sub longest {
     print $h "#define FEATURE_BUNDLE_$key	", $count++, "\n";
 }
 
+# I think that the test for (CURRENT_HINTS & HINT_LOCALIZE_HH) can actually
```
"I" in a comment in committed code is rather unclear. Either put your name, or add a signature line to the comment to explain who you are who are thinking this.

demerphq replied:
Ill fix that.

demerphq replied:
Fixed.

leonerd reviewed (commented):
Overall the change looks good.

The change also ensures that generated COPs contain the `->cop_features` copied from the compiling state. That's a key useful change that we're after now, but the commit message doesn't really mention that part. It might be useful to mention that specifically as well so it's clearer we intend to add that too.

Or maybe split it into a separate commit to draw more attention to it.

demerphq committed:
feature.h - simplify expression CURRENT_HINTS expression

PL_hints expands to PL_compiling.cop_hints. So there is no
point in doing

    PL_curcop == &PL_compiling ? PL_hints : PL_curcop->cop_hints

as it resolves to doing:

    PL_curcop->cop_hints

but with a comparison, which likely kills any speed advantage
the direct access might provide.

The old define is left around to validate that the two are the same
if needed.

leonerd commented on cop.h:
```diff
@@ -558,6 +558,8 @@ string C<p>, creating the package if necessary.
 #define CopHINTHASH_get(c)	((COPHH*)((c)->cop_hints_hash))
 #define CopHINTHASH_set(c,h)	((c)->cop_hints_hash = (h))
 
+#define CopFEATURES_setfrom(c, o) ((c)->cop_features = (o)->cop_features)
```
Tiniest of nitpics: isn't it traditional to use `dst` and `src` for these two arguments?

demerphq replied:
This seems in keeping with the style of the file, and is harmless, even though generally id agree with you that it would be more comprehensible.

demerphq replied:
Fixed.

demerphq commented:
@leonerd: I split out the logic copying the cop_features data by integrating the patch from @tonycoz which uses a macro for the copy.

I also added a patch which simplifies the CURRENT_HINTS macro, which was doing an unnecessary ternary to avoid indirection on the PL_curcop pointer.

demerphq commented:
I pushed some fixes to update the version mentioned in @nwc10 patch since it is now a bit out of date.

tonycoz commented:
> The assumption is that cop_features holds the bitmask of the currently enabled feature set, and hence that S_magic_sethint_feature() can incrementally update it as keys for features are set in/cleared from %^H.
> 
> However, this overlooks one problem - the default set of features enabled (keys) is not empty, the default for %^H _is_ empty, but as long as %^H is empty the internals knows that it is _not_ running with a custom feature set, so avoids the complexity of setting up %^H.

"The assumption is that cop_features holds the bitmask of the currently enabled feature set" - cop_features is intended to reflect the current state of the feature* keys in `%^H`, this is only matches the current set of enabled features when the feature bundle (in cop_hints) is set to FEATURE_BUNDLE_CUSTOM.

I thought this change might break feature.pm setting features when switching from a feature bundle to a custom set, but feature.pm clears every %^H feature, even those not currently set (hence setting cop_features to zero) before setting the feature flags for that bundle.

"but as long as %^H is empty" - cop_hints, not $^H is looked at here.  %^H isn't only used for features, so this wouldn't work anyway.

```
use v5.36; # bundle set to FEATURE_BUNDLE_535, cop_features/%^H ignored for features
use feature 'indirect'; # bundle set to FEATURE_BUNDLE_CUSTOM, cop_features has meaning, feature.pm sets appropriate $^H{feature...} keys
use v5.34; # bundle set to FEATURE_BUNDLE_527, cop_features/%^H have features set to 535 bundle features + indirect, but are ignored
```

So I think the set-default-cop_features change has no real effect, and may confuse someone looking at the behaviour of the code.

One thing we could do here is eliminate feature bundles entirely, since they I assume they originally existed to avoid a costly %^H lookup.  We no longer pay that cost, and it would save the (currently always compile-time) cost of checking for the bundle, and get 4 hints bits back.  This has some complexities though.

The copying of the feature bits from PL_compiling to the new state op is still needed and unrelated to my comments above.

demerphq commented:
Sorry, some of the terms used here are confusing me. When you say "default bundle" you mean the patch from @nwc10? So you think we should just drop @nwc10's patch and only keep yours and mine? I am fine with that. It looked reasonable to me, but if you are convinced otherwise im fine with that. I had the impression from @nwc10 and the patch that if we apply it we can remove some other logic aftewards. But im not fussed. I just want the cop_features copy done properly.

demerphq commented:
Ive removed the default features patch.
